Zend/zend_operators.c

```diff
@@ -2615,26 +2640,37 @@ ZEND_API zend_result ZEND_FASTCALL increment_function(zval *op1) /* {{{ */
 					default:
 						/* Perl style string increment */
 						increment_string(op1);
+						if (EG(exception)) {
+							return FAILURE;
+						}
 						break;
 				}
 			}
 			break;
 		case IS_FALSE:
 		case IS_TRUE:
-			/* Do nothing. */
+			zend_error(E_WARNING, "Increment on type bool has no effect, this will change in the next major version of PHP");
 			break;
 		case IS_REFERENCE:
 			op1 = Z_REFVAL_P(op1);
 			goto try_again;
-		case IS_OBJECT:
+		case IS_OBJECT: {
 			if (Z_OBJ_HANDLER_P(op1, do_operation)) {
 				zval op2;
 				ZVAL_LONG(&op2, 1);
 				if (Z_OBJ_HANDLER_P(op1, do_operation)(ZEND_ADD, op1, op1, &op2) == SUCCESS) {
 					return SUCCESS;
 				}
 			}
+			zval tmp;
+			if (Z_OBJ_HT_P(op1)->cast_object(Z_OBJ_P(op1), &tmp, _IS_NUMBER) == SUCCESS) {
+				ZEND_ASSERT(Z_TYPE(tmp) == IS_LONG || Z_TYPE(tmp) == IS_DOUBLE);
+				zval_ptr_dtor(op1);
```

Comment:
This may throw in `__destruct`. What are the correct semantics in this case? I think it's already correct, we should continue with the assignment and increment the casted value. Might be worth adding a test.

Reply:
This will require a custom object in ext/zend_test as no userland object will pass a ``_IS_NUMBER`` cast.

Reply:
True. I'm ok with leaving this untested.